### DISCUSSION TITLE
mesh.find_containing_cell DOES NOT work properly

### DISCUSSION THREAD

**Discussion by zixiao-yin** [pyvista/pyvista#6152] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
mesh.find_containing_cell()
```

### Description

Hey dear community,
I just found out that the function find_containing_cell() does not work properly.
It generates different results from point_cell_ids(). I’m not sure what causes this problem.
Please see the screenshots below.
Let’s say we have a point id of 6182. Using mesh.point_cell_ids(6182) I got a return of [137], which means the cell 137 contains this point 6182. I checked the cell 137. And yes, it indeed contains the point 6182.
Then the problem comes. When I try to use mesh.find_containing_cell(mesh.points[6183,:]) to locate the cell that contains point 6182, I got a return of 15, which indicates the cell 15 contains this point 6182. Then I also checked the cell 15. However, it does not actually contain the point 6182.
Anyone knows what causes this strange difference?
Thanks in advance!
Zixiao

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Thu May 30 10:26:46 2024 西欧夏令时

                OS : Windows (10 10.0.19045 SP0 Multiprocessor Free)
            CPU(s) : 8
           Machine : AMD64
      Architecture : 64bit
               RAM : 31.9 GiB
       Environment : Jupyter
        GPU Vendor : Intel
      GPU Renderer : Intel(R) UHD Graphics 620
       GPU Version : 4.5.0 - Build 30.0.101.1122
  MathText Support : False

  Python 3.10.4 | packaged by conda-forge | (main, Mar 24 2022, 17:32:50) [MSC
  v.1929 64 bit (AMD64)]

           pyvista : 0.43.8
               vtk : 9.3.0
             numpy : 1.22.3
        matplotlib : 3.8.2
            scooby : 0.10.0
             pooch : 1.8.0
            pillow : 9.1.0
           imageio : 2.34.0
           IPython : 8.3.0
        ipywidgets : 8.1.1
             scipy : 1.8.0
              tqdm : 4.66.1
        jupyterlab : 4.1.4
      nest_asyncio : 1.5.5
--------------------------------------------------------------------------------
```

### Screenshots

![image](https://github.com/pyvista/pyvista/assets/55424621/c40966fc-a326-446d-be89-6c654f2aa8f3)
![image](https://github.com/pyvista/pyvista/assets/55424621/2cd38f6e-6e10-4cc9-95dd-80a58513c1df)

**Comment by MatthewFlamm** [comment]
Can you cook up a reproducible example?  One potential problem is that you are asking to find which cell contains a nodal point of the mesh, assuming this isn't a poly line or poly vertex cell.  Given the large number of points per cell, I suspect it might be a polyline?

**Reply by zixiao-yin** [reply]
Yes, it is a polyline. Would that impact results? Please see attached mesh file. [mesh_test.zip](https://github.com/pyvista/pyvista/files/15500684/mesh_test.zip)

Hopefully you can reproduce the probelm through:
mesh.point_cell_ids(6182)
mesh.get_cell(137).point_ids
mesh.find_containing_cell(mesh.points[6182,:])
mesh.get_cell(15).point_ids

**Reply by MatthewFlamm** [reply, marked as answer]
`find_containing_cell` finds the cells that contain a point in space, not necessarily a point of the mesh.  One problem with using a point of the mesh is that point is usually on the edge, or probably corner, of multiple cells.  It isn't inside a single cell typically.  It could be that your point belongs to multiple cells.  I'm not sure if this is the problem yet.  I don't recommend doing this in general.  `find_containing_cell` is for arbitrary points in 3D space rather than nodal points of the mesh.

Specifically for polyline cells, you could also be choosing an internal point that defines the cell, not a cell edge defining point.  In theory this one should work based on the above. I think a minimal reproducible example is needed to determine what is going on to tease this out.  It could be that vtk is not accurately representing a complex polyline shape for this function.  It could also be that lines do not behave well in general as numerical issues will be severe here, i.e. you have an infinitely thin cell and are trying to decide whether there is a point inside it.

**Comment by user27182** [comment]
You could try using the extract_points filter instead. With the default options (include_cells=True and adjacent_cells=True) this should extract any cells of the mesh which contain the specified points. You can then access the vtkOriginalCellIds array to get the cell ids of the input mesh for those points. I.e.:

`cell_ids = mesh.extract_points(point_ids)['vtkOriginalCellIds']`

**Reply by zixiao-yin** [reply]
Yeah, cool! I jsut checked it out. mesh.extract_points outputs the same result as mesh.point_cell_ids. I guess as Matthew said, to find the id of an exact point (rather than an arbitrary point in the space), I should avoid using find_containing_cell, which seems to be more complex

**Reply by user27182** [reply]
Oh I see, you already had the result you were looking for with `mesh.point_cell_ids(6182)`, but the question was more about why `find_containing_cells` gave a different result. 

I'm glad the issue could be cleared up.
